discovery/xds/client.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+)
+
+var userAgent = fmt.Sprintf("Prometheus/%s", version.Version)
+
+// ResourceClient exposes the xDS protocol for a single resource type.
+// See https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#rest-json-polling-subscriptions
+type ResourceClient interface {
+	// ResourceTypeUrl is the type URL of the resource
+	ResourceTypeURL() string
+
+	// Server is the xDS Management server
+	Server() string
+
+	// Fetch requests the latest view of the entire resource state
+	// If no updates have been made since the last request, the response will be nil
+	Fetch(ctx context.Context) (*v3.DiscoveryResponse, error)
+
+	// ID returns the ID of the client that is sent to the xDS server
+	ID() string
+
+	// Close releases currently held resources
+	Close()
+}
+
+type HTTPResourceClient struct {
+	client *http.Client
+	config *HTTPResourceClientConfig
+
+	// endpoint is the fully-constructed xDS HTTP endpoint
```

Comment:
```suggestion
	// Endpoint is the fully-constructed xDS HTTP endpoint.
```

Reply:
`endpoint` in lowercase is the name of the field here, should this be kept as is but add the `.` at the end?